build_system/clerk_driver.ml

```diff
@@ -157,366 +176,507 @@ type expected_output_descr = {
-            ]))
-      in
-      let file_str = really_input_string ic (in_channel_length ic) in
-      let test_declarations = Re.all test_rex file_str in
-      List.map
-        (fun groups ->
-          let id =
-            match Re.Group.get_opt groups 1 with
-            | Some x -> x
-            | None ->
-              Message.raise_error
-                "A test declaration is missing its identifier in the file %s"
-                file
-          in
-          let cmd = Re.Group.get groups 2 in
-          { tested_filename = file; output_dir; cmd; id })
-        test_declarations)
-        [@ocamlformat "disable"]
+let scan_tree (dir : File.t) : catala_build_item Seq.t =
+  File.scan_tree
+    (fun f ->
+      match get_lang f with
+      | None -> None
+      | Some lang -> Some (scan_catala_file f lang))
+    dir
 
-(** Var references used in the Clerk file *)
-module Var = struct
-  let tested_file = Nj.Expr.Var "tested_file"
-  let catala_cmd = Nj.Expr.Var "catala_cmd"
-  let expected_output = Nj.Expr.Var "expected_output"
-  let test_file_or_folder = Nj.Expr.Var "test_file_or_folder"
-
-  let name = function
-    | Nj.Expr.Var n -> n
-    | _ -> invalid_arg "Clerk_driver.Var.name"
-end
+(** {1 System analysis} *)
 
-let pipe_diff_cmd =
-  let open Nj.Expr in
-  let has_patdiff = Sys.command "type patdiff >/dev/null 2>&1" = 0 in
-  if has_patdiff then
-    Seq
-      [
-        Lit "|";
-        Lit "patdiff";
-        Seq [Lit "-alt-new"; Lit "current-output"];
-        Var.tested_file;
-        Lit "/dev/stdin";
-      ]
-  else Seq [Lit "| diff -u -b --color"; Var.tested_file; Lit "-"]
-
-let inline_test_rule catala_exe catala_opts =
-  let open Nj.Expr in
-  Nj.Rule.make "inline_tests"
-    ~command:
-      (Seq
-         [
-           Lit Sys.argv.(0);
-           Lit "runtest";
-           Lit ("--exe=" ^ catala_exe);
-           Lit ("--catala-opts=\"" ^ String.escaped catala_opts ^ "\"");
-           Var.tested_file;
-           pipe_diff_cmd;
-         ])
-    ~description:(Seq [Lit "INLINE TESTS of file"; Var.tested_file])
-
-let inline_reset_rule catala_exe catala_opts =
-  let open Nj.Expr in
-  Nj.Rule.make "inline_tests_reset"
-    ~command:
-      (Seq
-         [
-           Lit Sys.argv.(0);
-           Lit "runtest";
-           Lit ("--exe=" ^ catala_exe);
-           Lit ("--catala-opts=" ^ catala_opts);
-           Lit "--reset";
-           Var.tested_file;
-         ])
-    ~description:(Seq [Lit "RESET INLINE TESTS of file"; Var.tested_file])
-
-let add_reset_rules_aux
-    ~(redirect : string)
-    ~(rule_name : string)
-    (catala_exe_opts : string)
-    (rules : Rule.t Nj.RuleMap.t) : Rule.t Nj.RuleMap.t =
-  let reset_common_cmd_exprs =
-    Nj.Expr.
+(** Some functions that poll the surrounding systems (think [./configure]) *)
+module Poll = struct
+
+  (** Scans for a parent directory being the root of the Catala source repo *)
+  let catala_project_root: File.t option Lazy.t =
```

Comment:
1. the fine-tuning is disabled if not running from the catala tree ; the issue I was trying to solve is that, in normal operation, all libraries should be installed inside `<prefix>/lib/catala` ; but when running the catala tests themselves it would be pretty inconvenient to have to install the full project every time, and you would always have the risk of inconsistency between the installed artifacts and what you actually wanted to test.
  Another possible workaround, more error-prone but simpler, would be to simply enforce that through the environment when running the catala tests.
2. well we don't want all these to unconditionnally run at program initialisation (imagine they fail and it breaks even `--help` ? You wouldn't get proper handling of the debug flags, etc.) — and we also certainly don't want to run them multiple times if queried multiple times. So that's what `lazy` is for :D